### PR TITLE
WIP: Auto rules tests

### PR DESCRIPTION
**Pull request by Ambrevar**
# Description

This adds the long awaited auto-rules test.  Since it's currently broken (see #2895), we are in urgent need of a test suite to validate the many edge cases.

Should help with #2895.

# Discussion

Just a draft for now, there is much more to test.

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.

- [x] Git hygiene:
  - I have pulled from master before submitting this PR
  - There are no merge conflicts.
- [x] I've added the new dependencies as:
  - ASDF dependencies,
  - Git submodules,
    ```sh
	cd /path/to/nyxt/checkout
    git submodule add https://gitlab.common-lisp.net/nyxt/py-configparser _build/py-configparser
    ```
  - and Guix dependencies.
- [x] My code follows the style guidelines for Common Lisp code. See:
  - [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf)
  - [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml)
- [ ] I have performed a self-review of my own code.
- [ ] My code has been reviewed by at least one peer.  (The peer review to approve a PR counts.  The reviewer must download and test the code.)
- [ ] Documentation:
  - All my code has docstrings and `:documentation`s written in the aforementioned style.  (It's OK to skip the docstring for really trivial parts.)
  - I have updated the existing documentation to match my changes.
  - I have commented my code in hard-to-understand areas.
  - I have updated the `changelog.lisp` with my changes if it's anything user-facing (new features, important bug fix, compatibility breakage).
  - I have added a `migration.lisp` entry for all compatibility-breaking changes.
  - (If this changes something about the features showcased on Nyxt website) I have these changes described in the new/existing article at Nyxt website or will notify one of maintainters to do so.
- [ ] Compilation and tests:
  - My changes generate no new warnings.
  - I have added tests that prove my fix is effective or that my feature works.  (If possible.)
  - I ran the tests locally (`(asdf:test-system :nyxt)` and `(asdf:test-system :nyxt/gi-gtk)`) and they pass.

### CODE REVIEW

**Comment by Ambrevar** [comment]
@aartaka The `auto-rules-remembered-modes` last 2 assertions are failing.  In my understanding, this is an auto-rules bug.  Could you look into it?

**Comment by aartaka** [comment]
> @aartaka The `auto-rules-remembered-modes` last 2 assertions are failing. In my understanding, this is an auto-rules bug. Could you look into it?

Yeah, that looks like no-rule -> no-rule transition, which currently unconditionally restores the last active modes over the enabled ones.

**Comment by Ambrevar** [comment]
Commits https://github.com/atlas-engineer/nyxt/pull/2906/commits/f22395bd4441c0414aee15c18d1c51cc9eb88d59 and https://github.com/atlas-engineer/nyxt/pull/2906/commits/e877455acd0007cb39149c7c545208749512d7f1 fix an issue with `cl:pathname` in `quri:uri` path.

I believe however that the fix belongs upstream.  See https://github.com/fukamachi/quri/issues/78.

**Comment by aartaka** [comment]
I've fixed it, I think. But I'll have to test it more.

In the meantime, here's a full matrix of cases we have to test (with combinations that we currently have marked with names of tests including them):
```
| transition\rule    | exclude | include              | exclude+include |
| rule -> rule       |         |                      |                 |
| no rule -> no rule |         | remembered-modes + 1 |                 |
| rule -> no rule    |         | navigation... + 1    |                 |
| no rule -> rule    |         | basic                |                 |
```

But this framework is insufficient to describe all the scenarios. Additional situations to test too (and that current tests more or less manage):
- Reloading the page (rule, no rule).
- Manual mode toggling (same matrix).

**Comment by aadcg** [comment]
Do we really need the renderer to test `auto-rules`?

**Comment by Ambrevar** [comment]
@aartaka Great work with the matrix, I'll get down to it.

**Comment by aadcg** [comment]
@Ambrevar makes sense, thanks for the clarification!

**Comment by aartaka** [comment]
A way to cut some corners: test all the page transitions with just one rule (like we are doing now), and then test all the rule types with some typical use scenario (like no-rule -> rule -> no-rule transition). It's not exactly bullet-proof, but possible to implement fast and covers the most matrix space with least effort spent.

File: source/auto-rules.lisp [hunk 1]
```diff
@@ -157,8 +157,10 @@ ARGS as in make-instance of `auto-rule'."
         (disable-modes* modes buffer)))))
 
 (defun can-save-last-active-modes-p (buffer url)
-  (or (null (last-active-modes-url buffer))
-      (not (quri:uri= url (last-active-modes-url buffer)))))
+  (and url
+       (or (null (last-active-modes-url buffer))
+           (not (quri:uri= url (last-active-modes-url buffer)))
+           (not (matching-auto-rules (previous-url buffer) buffer)))))
```
Comment (Ambrevar) [inline source/auto-rules.lisp:163]: Can you add a docstring to explain what this does?

**Comment by Ambrevar** [comment]
Commit 294b2c207abb69da385a0bf88309c6f7738efe16 with mode exclusion fails.  @aartaka Can you look into it?

**Comment by aartaka** [comment]
> * the logic behind `top-level-p` in `gtk.lisp`

See the new documentation in `toplevel-p` slot of `request-data`.

> * `save-last-active-modes`
> * `reapply-last-active-modes` (which are the "last" active modes?  Consider the case of matched rule and unmatched rule)
> * `remember-on-mode-toggle`: Why is `save-last-active-modes` called if `prompt-on-mode-toggle-p` is NIL and not in the other branch?

See the updated docs and `last-active-modes` slot documentation.

> * `add-modes-to-auto-rules`: Docstring including an example usage that involves `url-infer-match` and `rememberable-of`.

I'm wondering if `url-infer-match` should actually be part of `add-modes-to-auto-rules`...

**Comment by aartaka** [comment]
This is extremely weird. Test fails locally, but the interactive session works just fine (・_・ヾ

**Comment by aadcg** [comment]
What's the status here?

**Comment by aartaka** [comment]
Partially fixed on master, but needs more tests (and, supposedly, more fixes).

**Comment by aartaka** [comment]
> What's the status here?

Partially fixed on master, but needs more tests (and, supposedly, more fixes).

**Comment by aadcg** [comment]
I think this is not the right way of handling communication in a collaborative project.

If you have merged a subset of commits featured in this PR, you should have informed about it _beforehand_ in this thread.

Then you should inform the parties involved about the future of this PR. Should it be closed, rebased, etc? Loose ends are enemies of a successful project. 

---

> Partially fixed on master, but needs more tests (and, supposedly, more fixes).

What should we do about this PR?

**Comment by Ambrevar** [comment]
We need to finish it, as the tests (about mode exclusion) do not all pass.  See comment https://github.com/atlas-engineer/nyxt/pull/2906#issuecomment-1527576403.

**Comment by jmercouris** [comment]
Please reopen when ready.
